pretty printing for quotients, graded and localized multivariate polynomial rings

simonbrandhorst opened the pull request:
This pr should probably wait until the changes from nemo and abstract algebra propagate.

afkafkafk13 commented on src/Rings/MPolyQuo.jl:
```diff
@@ -36,11 +36,22 @@ function groebner_assure(r::MPolyQuoRing)
+function Base.show(io::IO, ::MIME"text/plain", Q::MPolyQuoRing)
+   println(io, "Quotient")
+   println(io, "  of ", base_ring(Q))
+   print(io, "  by Ideal with " , ngens(modulus(Q)), "generators")
```
The key information of the quotient ring is the base_ring and the modulus. 

I agree that in standard view only the number of generators is printed, but there should be an even more detailed view available, which prints the modulus (and not in supercompact printing of an ideal). Think of a user having defined several quotients of the same base_ring and trying to pick the right one for the next computation.

simonbrandhorst replied:
Some more information on the modulus is probably desirable in standard view. 
There will be no more detailed view than standard view. In this case you have to ask the quotient ring for the information that you desire.

afkafkafk13 replied:
Indeed, this also solves the dilemma of how much information is necessary.

simonbrandhorst commented on src/Rings/localization_interface.jl:
```diff
@@ -383,11 +383,23 @@ canonical_unit(f::LocRingElemType) where {LocRingElemType<:AbsLocalizedRingElem}
+  else
+    print(io, "Localization of ", Lowercase(), base_ring(W))
+    print(pretty(IOContext(io, :supercompact =>true)), " at ", Lowercase(), inverted_set(W))
+  end
```
@thofma this is it.

thofma commented on src/Rings/mpolyquo-localizations.jl:
```diff
@@ -1634,18 +1724,37 @@ function ideal(
+  if n > 0
+    print(io, Dedent())
+    println(io, "with ", ItemQuantity(ngens(I),"generator"))
+    print(io, Indent())
```
Missing a `Dedent()`

thofma commented on src/Rings/localization_interface.jl:
```diff
@@ -383,11 +393,24 @@ canonical_unit(f::LocRingElemType) where {LocRingElemType<:AbsLocalizedRingElem}
+    print(io, "Localized ring")
+  else
+    print(io, "Localization of ", Lowercase(), base_ring(W))
+    print(IOContext(io, :supercompact =>true), " at ")
```
Probably want the context in the line below?

thofma commented on src/Rings/mpoly-localizations.jl:
```diff
@@ -92,9 +92,37 @@ Base.iterate(U::MPolyPowersOfElement, i::Int) = (i < length(U.a) ? (U.a[i+1], i+
+function Base.show(io::IO, ::MIME"text/plain", S::MPolyPowersOfElement)
+  io = pretty(io)
+  println(io, "Multiplicative subset")
+  print(io, Indent())
```
Missing a `Dedent()`

thofma commented on src/Rings/mpoly-localizations.jl:
```diff
@@ -92,9 +92,37 @@ Base.iterate(U::MPolyPowersOfElement, i::Int) = (i < length(U.a) ? (U.a[i+1], i+
+    if i!=length(denominators(S))
+      print(io, ", ")
+    end
+  end
```
Not too important, but all those could be simplified by writing `print(io, "["); join(io, denominators(S), ", "); print(io, "]")`

simonbrandhorst replied:
ah that is good to know. I will edit. Bad code tends to multiply.

thofma reviewed (commented):
A part from a few things, looks good from a technical point of view.

afkafkafk13 commented on docs/src/CommutativeAlgebra/GroebnerBases/groebner_bases.md:
```diff
@@ -71,10 +71,7 @@ julia> default_ordering(F)
-  x -> [1]
-  y -> [2]
-  z -> [3], MPolyDecRingElem{QQFieldElem, QQMPolyRingElem}[x, y, z])
+(Graded multivariate polynomial ring in 3 variables over QQ, MPolyDecRingElem{QQFieldElem, QQMPolyRingElem}[x, y, z])
```
minor nitpick: printing of the variables still needs improvement

simonbrandhorst replied:
how?

afkafkafk13 replied:
'... over QQ, variables [x,y,z]'

simonbrandhorst replied:
What we print here is a `tuple (Ring, Vector{Variables})  `  because the constructor also returns the variables.
Arrays always print their type. Since we do not own arrays (they come from `Base`), we cannot change this. 
And it might not be a good idea even if we could.

afkafkafk13 replied:
I see.
What struck me, was the enormous disbalance between the amount of detail printed for the first entry (nice and suitable for mathematicians) and the second entry (which easily takes more than a line before even opening the square bracket, if you are dealing with a graded ring over an algebraic extension of QQ).

lgoettgens replied:
Something very similar has been discussed in https://github.com/oscar-system/Oscar.jl/issues/2064, although without a result.

simonbrandhorst replied:
For (multivarite) rings a solution would be to return the generators as a tuple and not as an array. The tuple does not print its type.

simonbrandhorst replied:
.... but this is not type stable, so we are back to arrays.

afkafkafk13 commented on docs/src/CommutativeAlgebra/localizations.md:
```diff
@@ -421,7 +437,12 @@ julia> U = complement_of_point_ideal(R, [0, 0]);
+
+with 2 generators
+  3*x^2
+  4*y^3
```
newline problem?

afkafkafk13 commented on experimental/Schemes/BlowupMorphism.jl:
```diff
@@ -404,11 +404,11 @@ Spec of Multivariate polynomial ring in 3 variables over QQ
 =====================================
 Affine charts of the blown up scheme:
-Spec of Localization of Quotient of Multivariate polynomial ring in 3 variables over QQ by ideal(0, 0, 0) at the multiplicative set powers of QQMPolyRingElem[1]
+Spec of Localization of quotient of multivariate polynomial ring at multiplicative subset
```
I would suggest to be a bit more precise with the multiplicative set here, if this is not compact/supercompact: '... ring by inverting powers of an element' and suppress the multiplicative set completely in more compact printing. If we localize, we do it at a mult. closed subset, so the current information is redundant.

afkafkafk13 commented on src/AlgebraicGeometry/Schemes/AffineSchemes/Objects/Constructors.jl:
```diff
@@ -39,7 +39,7 @@ julia> I = ideal(R, [x]);
 
 julia> Spec(R, U)
-Spec of localization of Multivariate polynomial ring in 2 variables over QQ at the complement of ideal(x)
+Spec of Localization of multivariate polynomial ring in 2 variables over QQ at Complement of prime ideal
```
Do you really want this in capital letters? (Localization, Complement)

afkafkafk13 commented on src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Objects/Attributes.jl:
```diff
@@ -216,10 +211,7 @@ julia> R, (u, v) = QQ["u", "v"];
-  x -> [1]
-  y -> [1]
-  z -> [1], MPolyDecRingElem{MPolyQuoRingElem{QQMPolyRingElem}, AbstractAlgebra.Generic.MPoly{MPolyQuoRingElem{QQMPolyRingElem}}}[x, y, z])
+(Graded multivariate polynomial ring in 3 variables over Quotient of multivariate polynomial ring, MPolyDecRingElem{MPolyQuoRingElem{QQMPolyRingElem}, AbstractAlgebra.Generic.MPoly{MPolyQuoRingElem{QQMPolyRingElem}}}[x, y, z])
```
minor nitpick: second and third argument still need pretty printing.

afkafkafk13 commented on src/Combinatorics/Matroids/matroid_strata_grassmannian.jl:
```diff
@@ -15,7 +15,14 @@ julia> A # The coordinate matrix with entries in the polynomial ring `R`.
-Localization of Quotient of Multivariate polynomial ring in 9 variables over GF(2) by ideal(x[2, 3]*x[3, 4] + x[3, 3]*x[2, 4], x[1, 2]*x[3, 4] + x[3, 2]*x[1, 4], x[1, 1]*x[2, 4] + x[2, 1]*x[1, 4], x[1, 1]*x[3, 2]*x[2, 3] + x[2, 1]*x[1, 2]*x[3, 3]) at the multiplicative set powers of fpMPolyRingElem[x[3, 3]*x[1, 4], x[1, 1]*x[2, 3]*x[3, 4] + x[1, 1]*x[3, 3]*x[2, 4] + x[2, 1]*x[3, 3]*x[1, 4], x[2,  ...
+Localization
+  of quotient of multivariate polynomial ring by ideal with 4 generators
+  at powers of [x[3, 3]*x[1, 4], x[1, 1]*x[2, 3]*x[3, 4] + x[1, 1]*x[3, 3]*x[2, 4] + x[2, 1]*x[3, 3]*
```
maybe shorten the last part 'inverting all powers/products of 21 elements'

afkafkafk13 commented on src/Rings/PBWAlgebra.jl:
```diff
@@ -1289,7 +1289,7 @@ julia> eliminate(I, [1, 2 ,3])
 
 julia> try eliminate(I, [z, a]); catch e; e; end
-ErrorException("no elimination is possible: subalgebra is not admissible")
+
```
Why did the error message disappear? Its printing seems to have been part of the doctest on purpose before. Please investigate to make sure not to break a test here.

afkafkafk13 commented on src/Rings/PBWAlgebra.jl:
```diff
@@ -1306,7 +1306,7 @@ julia> I = left_ideal(A, [p, q])
 
 julia> try eliminate(I, [q]); catch e; e; end   # in fact, no elimination ordering exists
-ErrorException("could not find elimination ordering")
+
```
see above.

afkafkafk13 commented on src/Rings/localization_interface.jl:
```diff
@@ -169,7 +179,7 @@ Domain:
 Codomain:
 =========
-localization of Multivariate polynomial ring in 3 variables over QQ at the complement of ideal(x)
+Localization of multivariate polynomial ring in 3 variables over QQ at Complement of prime ideal
```
Are all changes in the use of capital letters here intentional?

simonbrandhorst replied:
Partly. But this map type comes from hecke and is not touched here. In the transition phase things will be inconsistent.

thofma replied:
It should just print the codomain using `print(io, codomain(...))`, so it should work?

afkafkafk13 commented on src/Rings/localization_interface.jl:
```diff
@@ -383,11 +393,24 @@ canonical_unit(f::LocRingElemType) where {LocRingElemType<:AbsLocalizedRingElem}
+  println(io, "Localization", Indent())
+  println(io, "of ", Lowercase(), base_ring(W))
+  print(io, "at ")
+  print(io, Lowercase(), inverted_set(W))
```
pretty printing of inverted_set is ok for complements of maximal and prime ideals, but still needs changes for powers/products of elements.

afkafkafk13 commented on src/Rings/mpoly-localizations.jl:
```diff
@@ -92,9 +92,28 @@ Base.iterate(U::MPolyPowersOfElement, i::Int) = (i < length(U.a) ? (U.a[i+1], i+
+  print(io, "given by the powers of ")
+  print(io, "[")
+  join(io, denominators(S), ", ")
+  print(io, "]")
```
This can get really lengthy. May take the same route as for ideals: at powers/products of n elements

afkafkafk13 commented on src/Rings/mpoly-localizations.jl:
```diff
@@ -92,9 +92,28 @@ Base.iterate(U::MPolyPowersOfElement, i::Int) = (i < length(U.a) ? (U.a[i+1], i+
-  print(io, "powers of ")
-  print(io, denominators(S))
+  io = pretty(io)
+  if get(io, :supercompact, false)
```
In supercompact, either print nothing at all or something stating that we are dealing with powers/products of elements.
The 'multiplicative set' should be left for multiple localizations, where we see more general multiplicative sets (unless we simplify them, which is not even on our agenda...)

afkafkafk13 reviewed (commented):
I went through all the output and wrote at least one comment for each of the small nitpicks I have. Overall there are the following central points:
- better pretty printing mult. set inverting powers of elements
- something changed in try - catch doctests -- is this intentional
- some supercompact suggestions

simonbrandhorst committed:
show functions for graded quotient and localized MPolyRings

simonbrandhorst committed:
import pretty printing tool from AbstractAlgebra

simonbrandhorst committed:
fix doctests

afkafkafk13 commented on experimental/FTheoryTools/src/TateModels/constructors.jl:
```diff
@@ -30,16 +45,68 @@ Normal toric variety
+   @ Base ./intfuncs.jl:340
+ [4] top-level scope
+   @ none:1
 
```
This commit obviously broke something in the FTheory code of Martin and Andrew......

simonbrandhorst replied:
I was still messing with my git. The new branch should be okay now. I cleaned up the history.

simonbrandhorst commented on src/Rings/mpoly-localizations.jl:
```diff
@@ -2428,18 +2657,81 @@ function MPolyLocalizedRingHom(
+julia> PSI = hom(TL, RQL, RQL.([x]))
+Ring homomorphism
+  from localization of multivariate polynomial ring in 1 variable over QQ at complement of maximal ideal
+  to   Localization of quotient of multivariate polynomial ring at complement of maximal ideal
```
@thofma  the `Localization` should be lowercase ?

simonbrandhorst committed:
bump AA to 0.30.04 and remove some old printing tests

afkafkafk13 commented on src/Rings/mpoly-localizations.jl:
```diff
@@ -2489,9 +2781,21 @@ julia> UT = complement_of_point_ideal(T, [0]);
+julia> PHI = hom(RQL, TL, TL.([t, t^2, t^3]))
+Ring homomorphism
+  from Localization of quotient of multivariate polynomial ring at complement of maximal ideal
+  to   localization of multivariate polynomial ring in 1 variable over QQ at complement of maximal ideal
```
for MPolyQuo the localizations start with a capital letter, for MPolyRing with a lowercase letter. Probably there is still a typo in the MPolyQuo-case somewhere.

simonbrandhorst replied:
there is still something fishy in our printing system. @thofma  is investigating

afkafkafk13 reviewed (commented):
Looks good to me.
Remember to change the respective testset in mpoly-localizations.jl to agree with the new output format.

simonbrandhorst committed:
remove some more old printing tests

simonbrandhorst commented:
> Looks good to me. Remember to change the respective testset in mpoly-localizations.jl to agree with the new output format.

I removed these old tests and added some doctests instead.

afkafkafk13 reviewed (approved):
Now this is a clear approval.

Sideremark: Finding and solving the 'something fishy' in the printing system is related to this PR, but not part of the changes.

thofma committed:
Bump AA version

thofma committed:
some further fixes
